dvc/repo/add.py

```diff
@@ -55,6 +55,10 @@ def get_or_create_stage(
     path, wdir, out = resolve_paths(repo, target, always_local=to_remote and not out)
 
     try:
+        # How best to disable this line? With Skip Graph Checks Flag?
+        repo._skip_graph_checks = True
+        if getattr(repo, "_skip_graph_checks", False):
+            print("todo: skip graph checks")
```

Comment:
> I don't have a strong understanding of what a "Stage" is.

Internally, everything is a `Stage` in dvc.

What you want to do here is similar to following:

```python
if not skip_graph_checks:
   try:
      ...  # existing inner block inside try
   except OutputNotFoundError:
     continue

...  # existing block from except
```


We might want to add a warning in case of `skip_graph_checks=True` saying that:

>  partial or virtual add does not work when --skip-graph-checks are enabled.

Reply:
Is the "graph" referring to a graph of stages where edges represent dependencies between stages? Is the essence of what I want an optimization for the case where the graph is a disconnected set of nodes?

Reply:
I'm still unsure what the right thing to do in `get_or_create_stage` when this optimization is enabled. From what I understand we are attempting to lookup a "Stage" in `repo.find_outs_by_path` and that corresponds to a ".dvc" file? Can I use .dvc file and Stage interchangeably? I've been referring to them as "Sidecars", but I would like to understand the architecture of the repo better.

In `find_outs_by_path`, it calls `self.index.outs_graph`, which is one of the big time sinks, and that's what I'd like to avoid if possible. Is there an alternative way I can create the `dvc.output.Output` object when the "path" refers to a directory or file that corresponds to an existing or soon-to-be-created ".dvc" file?

As it stands my current implementation falls back to the case where it assumes the file is always a new .dvc file. But this optimization should apply in the case where the entire directory or file is tracked by the .dvc file.  Guidance on what to do here would be helpful. 

I'm also thinking that if there is any way to only crawl part of the graph without creating the entire thing, then perhaps every use-case can be made more efficient without needing to add a special case? (Although in the short-term I do just want to add the special case and move on). 

Does it make sense to add docstrings to some of these functions? It would be nice there were docs explaining a bit about each argument and in which cases it might be specified, and perhaps points to a reference if it uses terminology the user is expected to be familiar with?

Something else that would be incredibly useful to have are doctests that would give me a quick entrypoint into any function. However, for things like DVC, that would require some concise helper function to setup a demo repo so the doctests don't get ridiculous. I happen to have [some code](https://gitlab.kitware.com/computer-vision/simple_dvc/-/blob/main/simple_dvc/demo.py) that does this. I've been working on a DVC helper module called simple_dvc that improves my quality of life. An example that illustrates can be seen in the ["request" function](https://gitlab.kitware.com/computer-vision/simple_dvc/-/blob/main/simple_dvc/api.py#L378). Is there any interest in porting any of that work here?

I've pushed up a new commit that adds some of these docstrings, with questions where I don't know the answer. If someone wants to take a look and comment on what I've written / had questions about, perhaps we can consolidate that knowledge into the docstrings and make this codebase more approachable.

Reply:
> As it stands my current implementation falls back to the case where it assumes the file is always a new .dvc file. But this optimization should apply in the case where the entire directory or file is tracked by the .dvc file. Guidance on what to do here would be helpful.

I think the approach suggested by @skshetry [above](https://github.com/iterative/dvc/pull/10425#discussion_r1599857138) seems reasonable. Yes, it will fall back to creatin a new .dvc file. I think this is why we will have to add warnings. I don't see how we can do better without traversing the graph.